be/src/serde/protobuf_serde.cpp

```diff
@@ -11,16 +11,84 @@
 
 namespace starrocks::serde {
 
-int64_t ProtobufChunkSerde::max_serialized_size(const vectorized::Chunk& chunk) {
+EncodeContext::EncodeContext(const int col_num, const int encode_level)
+        : _session_encode_level(encode_level), _times(0) {
+    for (auto i = 0; i < col_num; ++i) {
+        _column_encode_level.emplace_back(_session_encode_level);
+        _raw_bytes.emplace_back(0);
+        _encoded_bytes.emplace_back(0);
+    }
+    DCHECK(_session_encode_level != 0);
+    // the lowest bit is set and other bits are not zero, then enable adjust.
+    if (_session_encode_level & 1 && (_session_encode_level >> 1)) {
+        _enable_adjust = true;
+    }
+}
+
+void EncodeContext::update(const int col_id, uint64_t mem_bytes, uint64_t encode_byte) {
+    DCHECK(_session_encode_level != 0);
+    if (!_enable_adjust) {
+        return;
+    }
+    // decide to encode or not by the encoding ratio of the first EncodeSamplingNum of every _frequency chunks
+    if (_times % _frequency < EncodeSamplingNum) {
+        _raw_bytes[col_id] += mem_bytes;
+        _encoded_bytes[col_id] += encode_byte;
+    }
+}
+
+// if encode ratio < EncodeRatioLimit, encode it, otherwise not.
+void EncodeContext::_adjust(const int col_id) {
+    auto old_level = _column_encode_level[col_id];
+    if (_encoded_bytes[col_id] < _raw_bytes[col_id] * EncodeRatioLimit) {
+        _column_encode_level[col_id] = _session_encode_level;
+    } else {
+        _column_encode_level[col_id] = 0;
+    }
+    if (old_level != _column_encode_level[col_id] || _session_encode_level < -1) {
+        LOG(WARNING) << "Old encode level " << old_level << " is changed to " << _column_encode_level[col_id]
```

Comment:
Is it more appropriate to use debug log here?

Reply:
I do not find `LOG(DEBUG)` in SR, can we use it?

Reply:
> I do not find `LOG(DEBUG)` in SR, can we use it?

maybe you need to use VLOG